### DISCUSSION TITLE
What is your current git workflow, until support is added to helix?

### DISCUSSION THREAD

**Discussion by cd-a** [helix-editor/helix#4193] [Q&A]
Following https://github.com/helix-editor/helix/issues/227, git support is not yet implemented.

Until this is the case, in whatever form, for everyone using helix at the moment: 

**What's your current git workflow using helix?**

Do you switch terminals to lazygit, use another git viewer, etc?

**Comment by jcornaz** [comment]
Personally, the only thing I want from helix is the git gutters (#3890), and I use my own fork of helix with that PR included.

To perform git operations, I simply use `git` in another terminal instance (I am using i3). And I don't want anything else. I tried `lazygit` for a while but given I already know bare git well, and I already have my own personal `fish` scripts to support my workflow, I find `lazygit` to be less comfortable.

**Reply by cd-a** [reply]
Interesting. I can't live without a proper git wrapper like Fugitive anymore. I find it so convenient

**Comment by David-Else** [comment]
The only git features I want in Helix are:

1. Git gutters
2. Having the name of the branch on the status line

I use Lazygit and find it amazing, it is by far the best GUI for Git I have seen. I think adding anything else to Helix would be bloat, all the work is done in Lazygit, there is no point in trying to compete.

I use Kitty as the terminal window manager and have Helix and Lazygit each in a window in their own tab. I use this layout:

`kitty.conf`
```
enabled_layouts       stack,tall
```
Then I can toggle Lazygit and Helix being next to each other or taking the full screen using `Ctrl-shift` `[` `]` `l`, It works perfectly :)

**Reply by cd-a** [reply]
Actually very interesting to see other people's view on that.

Up until now, I never thought I'd want something else than fugitive or magit built inside the editor. But you have a point, those apps are built for git and git only, and doing it well. 

Maybe best if helix focused on other features first before (eventually or maybe) tackling this.

**Reply by leungbk** [reply]
One nice thing about Fugitive and Magit is that text edits such as deletion of a hunk, or overwrites of uncommitted stuff via `:Gread`, through their respective interfaces, get registered into the Vim/Emacs undo history, allowing users to later undo their changes using their editor's default undo command. I would not expect Git frontends existing outside the editor to do this.

Emacs also allows you to bookmark Magit buffers and create org links to Magit and Forge buffers. Having editor-level support permits a lot of serendipitous synergies.

**Comment by dariooddenino** [comment]
I've always done everything in magit. I've been trying a different approach to not be tied to emacs.

I'm using commitizen to create commits and gitkraken to handle more complex stuff and interactions with github.

I guess the things I'm missing are basic integration like gutters, blame and yeah current branch.

**Comment by dimas-cyriaco** [comment]
> Do you switch terminals to lazygit

This.

**Comment by dead10ck** [comment]
Maybe I'm just used to it, but what's wrong with just `git` at the command line? That's pretty much always how I've used git. I also use `tig` for a better visualization of the tree and as a convenient way to invoke `git diff`, but otherwise the command line for everything else.

**Comment by ksandvik** [comment]
I've used and will use git command line tools as I know what those produce, having a UI wrapper between I don't control is of lesser interest, even if it looks pretty. But if there was a way to issue git commands (or any command such as cargo commands) from the :sh status line and the output was sent to a newly create horizontal temporary buffer shown, that would be nice.

**Comment by superDuperCyberTechno** [comment]
I use tmux so another terminal is always less than a second away. From there it's just the Git CLI.

The only thing I miss often enough to still use Vim is the Git gutter.

**Comment by goyalyashpal** [comment]
Luke Pighetti shared a method in his videos on helix where his terminals supported sending the helix in background, activating the CLI, then reactivating helix back.

if i get that in my terminal (wezterm on os win 10), then i'll have everything i need...

so, it don't need pretty much anything from helix for git at this point... 
yeah, having some alternative to vscodium for diff-ing (that not necessarily has to be helix) will be cool.

**Comment by terrybrash** [comment]
Missing Git features that block me from switching to Helix:
1. Staging and reverting selected lines (or hunks)
2. Side-by-side diff viewer

**Comment by James-Firth** [comment]
Diff comparisons and merging I use Sublime Merge.

The only git-related feature I would like is the ability to see some git blame information on all the lines for investigation/knowing who to ask questions

**Comment by acidindustries** [comment]
For those on Windows, I use Windows Terminal with LazyGit.

This is the config to start lazygit in a new split pane.

```toml
C-g = ":sh wt -w '0' split-pane -d . lazygit"

```

**Comment by matta** [comment]
For Emacs refugees that miss Magit, I recommend https://github.com/altsem/gitu. It has a very similar UI, lets you stage hunks easily down to the line level, etc.
